src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,368 @@
+                    resume_unwind(e);
+                }
+            }
+        }
+    }
+}
+
+/// Trait that implements ``read_ahead`` a method for converting ``self`` to a
+/// ``ChunkedReadAheadIterator``.
+#[allow(clippy::module_name_repetitions)]
+pub trait IntoChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    /// Converts a struct into a ``ChunkedReadAheadIterator``, with chunks of size `chunk_size`
+    /// and a read ahead buffer of `num_chunk_buffer_size` chunks.
+    fn read_ahead(
+        self,
+        chunk_size: usize,
+        num_chunk_buffer_size: usize,
+    ) -> ChunkedReadAheadIterator<T>
+    where
+        Self: Send + 'static;
+}
+
+impl<I, T> IntoChunkedReadAheadIterator<T> for I
+where
+    T: Send + 'static,
+    I: Iterator<Item = T>,
+{
+    fn read_ahead(
+        self,
+        chunk_size: usize,
+        num_chunk_buffer_size: usize,
+    ) -> ChunkedReadAheadIterator<T>
+    where
+        Self: Send + 'static,
+    {
+        ChunkedReadAheadIterator::new(self, chunk_size, num_chunk_buffer_size)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::rstest;
+    use std::mem::drop;
+    use std::panic;
+    // use std::panic;
+    use std::thread::sleep;
+    use std::time::Duration;
+
+    #[rstest]
+    #[case(1)] // smallest possible
+    #[case(2)]
+    #[case(4)]
+    #[case(8)]
+    #[case(16)] // larger than the inner iterator
+    fn test_wrapping_empty_iter(#[case] chunk_size: usize) {
+        let test_vec: Vec<usize> = Vec::new();
+
+        let mut chunked_iter = test_vec.into_iter().read_ahead(chunk_size, 1); // TODO add into chunked readahead
+        assert_eq!(chunked_iter.next(), None);
+    }
+
+    #[rstest]
+    #[case(1)] // smallest possible
+    #[case(2)]
+    #[case(4)]
+    #[case(8)]
+    #[case(16)] // larger than the inner iterator
+    fn test_handle_large_iterator_and_low_chunk_size(#[case] chunk_size: usize) {
+        let test_vec: Vec<usize> = (0..1_000_000).into_iter().collect();
+        let test_vec2 = test_vec.clone();
+
+        let mut regular_iter = test_vec.into_iter();
+        let mut chunked_iter = test_vec2.into_iter().read_ahead(chunk_size, 1);
+
+        loop {
+            let i = regular_iter.next();
+            let j = chunked_iter.next();
+            assert_eq!(i, j);
+            if i.is_none() {
+                assert!(j.is_none());
+                break;
+            }
+        }
+    }
+
+    #[test]
+    fn test_low_bound_on_channel_for_blocking() {
+        let mut chunked_iter = (0..100_000).into_iter().read_ahead(8, 1);
+        for _ in 0..4 {
+            chunked_iter.next();
+        }
+        drop(chunked_iter);
+        let mut test_iter = vec![0usize, 1, 2, 3, 4, 5, 6, 7, 8, 9].into_iter();
+        for i in 0..20 {
+            let v = test_iter.next();
+            if i < 10 {
+                assert_eq!(v, Some(i));
+            } else {
+                assert_eq!(v, None);
+            }
+        }
```

Comment:
What does the second half of this test do?

Reply:
... This test appears to be a pair of copy paste error that are not doing what its supposed to. Let me fix.